drivers/at24cxxx/at24cxxx.c

```diff
@@ -0,0 +1,366 @@
+/**
+ * @ingroup     drivers_at24cxxx
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for the AT24C128 / AT24C256 EEPROM units.
+ *
+ * @}
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+
+#define ENABLE_DEBUG  (0)
+#include "debug.h"
+#include "assert.h"
+#include "xtimer.h"
+
+#include "include/at24cxxx_params.h"
+#include "include/at24cxxx_defines.h"
+#include "at24cxxx.h"
+
+/*
+    page 0: 0-63
+    page 1: 64-127
+    page 2: 128-191
+    page 3: 192-255
+    page 4: 256-319
+    page 5: 320-383
+    page 6: 384-447
+    page 7: 448-511
+    page 8: 512-575
+    ...
+ */
+static inline void _position_to_address_words(uint8_t addr[AT24CXXX_ADDR_LEN],
+                                              uint32_t pos)
+{
+    uint16_t a = pos & 0xFFFF;
+
+    addr[0] = (a & 0xFF00) >> 8;
+    addr[1] = a & 0x00FF;
+}
+
+static int _read_byte(const at24cxxx_t *dev, uint32_t pos)
```

Comment:
I think you can replace this entire function with [`i2c_read_reg()`](https://riot-os.org/api/group__drivers__periph__i2c.html#gac4ab6e73e05b22c2da0c2f178da06d7b).

Reply:
How about `_read_byte()` calling `i2c_read_reg()`, while it also polls for the EEPROM to respond.

Reply:
So that after a write no delay and polling is added, but the next access (read/write) potentially takes a bit longer if thw previous write did no complete?

Sounds like the best solution, especially as the delay only is inserted when actually needed. (But keep in mind the potential hardware issue about polling for buggy I2C controllers I mentioned previously.)

Reply:
If I understood @MrKevinWeiss correctly, you could just start the next read/write and assume on a NACK that the last write cycle did not complete and retry. On an address NACK the H/W will not automatically read the next, right?

I personally like the proposal of @benpicco very much to sleep for 1ms and try 5 times to get a compromise between bus-sharing/power-consumption and access time.